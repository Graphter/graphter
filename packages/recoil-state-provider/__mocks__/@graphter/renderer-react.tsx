import { NodeRendererRegistration } from "@graphter/core";
import React from "react";

export const nodeRendererStore = {
  get: jest.fn()
}

export const pathConfigStore = {
  get: jest.fn(),
  set: jest.fn()
}

export const NodeValidationProvider = jest.fn(({ children }: any) => <div>{children}</div>)
export const NodeDataProvider = jest.fn(({ children }: any) => <div>{children}</div>)