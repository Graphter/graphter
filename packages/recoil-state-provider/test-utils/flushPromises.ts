import { act } from "@testing-library/react";

export default async function flushPromises() {
  return act(async () => {
    return new Promise(resolve => setImmediate(resolve));
  })
}