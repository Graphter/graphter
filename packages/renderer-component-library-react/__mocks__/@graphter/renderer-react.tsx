import React, { useState } from "react";

export const useArrayNodeData = jest.fn(() => {
  return {
    childIds: [],
    removeItem: jest.fn(),
    commitItem: jest.fn()
  }
})

export const useNodeData = jest.fn(() => {
  return [
    null,
    () => {}
  ]
})

export const useNodeValidation = jest.fn().mockResolvedValue([])

export const nodeRendererStore = {
  get: jest.fn((nodeType: string) => ({
    type: 'string',
    renderer: ({originalNodeData}: any) => {
      const [ value, setValue ] = useState(originalNodeData || '')
      return (
        <input value={value} onChange={e => setValue(e.target.value)} data-testid='string-renderer' />
    )
    }
  }))
}
export const createDefault = jest.fn()