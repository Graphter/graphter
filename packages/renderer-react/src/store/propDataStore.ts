import { atom, RecoilState } from "recoil";

const propDataModelMap: { [key: string]: RecoilState<any> } = {};
export default {
  get: (modelId: string, jsonPointer: string, defaultValue?: any) => {
    const key = `${modelId}-${jsonPointer}-c2c87429-dabf-4ea0-b2e1-6e7a6262bc11`
    let propDataState = propDataModelMap[key]
    if(!propDataState){
      if(defaultValue === undefined) throw new Error(`Can't create Recoil state without a default value`)
      propDataModelMap[key] = propDataState = atom({
        key,
        default: defaultValue
      })
    }
    return propDataState
  }
}