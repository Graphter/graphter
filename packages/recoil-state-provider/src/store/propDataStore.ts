/***
 * TODO: This module makes some pretty nasty data mutations. Fix that.
 */

import { atom, RecoilState } from "recoil";
import { PathSegment } from "@graphter/core";
import { nanoid } from "nanoid";

interface PropDataStateMeta {
  state: RecoilState<any>,
  deleted: boolean,
  committed: boolean,
  default: any
}

interface Node {
  id: string | null,
  meta: PropDataStateMeta | null,
  children: Array<Node> | null
}

const propStateNodeTree: Array<Node> = []

export const init = (
  path: Array<PathSegment>,
  data: any
): void => {
  const init = (path: Array<PathSegment>, data: any) => {
    if(Array.isArray(data)) {
      if(!has(path)) set(path, true, data.map(() => nanoid()))
      data.forEach((item, i) => init([ ...path, i ], item))
    } else if(typeof data === 'object') {
      if(!has(path)) set(path, true, data)
      Object.entries(data).forEach(([key, value]) => {
        init([ ...path, key ], value)
      })
    } else {
      if(!has(path)) set(path, true, data)
    }
  }
  init(path, data)
}

export const get = (
  path: Array<PathSegment>
): RecoilState<any> => {
  checkPathArg(path)
  const pathNode = getPathNode(path)
  if(!pathNode.meta) throw new Error(`Couldn't find state at ${path.join('/')}`)
  return pathNode.meta.state
}

export const getAll = (
  path: Array<PathSegment>
): Array<RecoilState<any>> => {
  checkPathArg(path)
  const pathNode = getPathNode(path)
  if(!pathNode.children) throw new Error(`Couldn't find child state at ${path.join('/')}`)
  return Array.from(pathNode.children.values())
    .filter(node => node.meta && node.meta.committed)
    .map(node => node.meta?.state as RecoilState<any>)
}

export const set = (
  path: Array<PathSegment>,
  committed: boolean,
  originalValue?: any
) => {
  checkPathArg(path)
  console.info(`Setting ${path.join('/')} = ${originalValue} [committed=${committed}]`)
  const node = getPathNode(path)
  node.meta = {
    state: atom({
      key: nanoid(),
      default: originalValue
    }),
    deleted: false,
    committed,
    default: originalValue
  }
}

/***
 * The "committed" flag in Graphter is used to demarcate draft tree additions. For example
 * the "add item" ui for a list before the "Add [+]" button is pressed.
 *
 * In future we'll likely have to integrate this function with node modification history or
 * versioning to allow us to track tree modifications as well as additions. For example a
 * renderer may want to allow changes and then provide a Save/Cancel button to commit or
 * discard the tree changes
 * @param path
 */
export const commitItem = (
  path: Array<PathSegment>
) => {
  checkPathArg(path)
  const node = getPathNode(path)
  if(!node.meta) throw new Error(`Node at ${path.join('/')} is missing meta. Should never happen`)
  node.meta.committed = true
}

export const has = (path: Array<string | number>): boolean => {
  checkPathArg(path)
  let node: Node | undefined = {
    id: null,
    meta: null,
    children: propStateNodeTree
  }
  for(let segment of path){
    if(typeof segment === 'string'){
      node = node.children?.find(childNode => childNode.id === segment)
    } else {
      node = node.children ? node.children[segment] : undefined
    }

    if(!node) return false
  }
  return true
}

export const remove = (path: Array<PathSegment>) => {
  checkPathArg(path)
  const clonedPath = [...path]
  const removeSegment = clonedPath.splice(-1, 1)[0]

  const parentNode = getPathNode(clonedPath)
  if(!parentNode) throw new Error(`Missing ancestor trying to set ${path.join('/')}`)
  if(!parentNode.children) throw new Error(`No children at ${path.join('/')} to remove`)
  const indexToRemove = typeof removeSegment === 'number' ?
    removeSegment :
    parentNode.children
      .findIndex(childNode => childNode.id !== removeSegment)
  parentNode.children.splice(indexToRemove, 1)
}

const getPathNode = (path: Array<PathSegment>) => {
  return path.reduce<Node>((a, c) => {
    if(!a.children) a.children = []
    let node = typeof c === 'string' ?
      a.children.find(node => node.id === c) :
      a.children[c]

    if(!node){
      node = {
        id: typeof c === 'string' ? c : null,
        meta: null,
        children: null
      }
      typeof c === 'string' ?
        a.children.push(node) :
        a.children[c] = node
    }
    return node
  }, { id: null, meta: null, children: propStateNodeTree })
}

function checkPathArg(path: Array<PathSegment>){
  if(!path || !path.length) throw new Error('Path of at least one segment is required')
}

export interface PropDataStore {
  init: (path: Array<PathSegment>, data: any) => void
  get: (path: Array<PathSegment>) => RecoilState<any>
  getAll: (path: Array<PathSegment>) => Array<RecoilState<any>>
  set: (
    path: Array<PathSegment>,
    committed: boolean,
    originalValue?: any
  ) => void
  commitItem: (path: Array<PathSegment>) => void
  has: (path: Array<string | number>) => boolean
  remove: (path: Array<PathSegment>) => void
}

export const propDataStore: PropDataStore = {
  init,
  get,
  getAll,
  set,
  commitItem,
  has,
  remove
}

export default PropDataStore