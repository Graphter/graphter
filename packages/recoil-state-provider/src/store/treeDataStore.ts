/**
 * WARNING: this file is being used directly in various places in the react renderer
 * which is against the principle that the renderer should be agnostic of state management choices.
 * TODO: refactor access to this module through a standardised interface
 */
import { RecoilValueReadOnly, selector } from "recoil";
import propDataStore from "./propDataStore";
import { PathSegment } from "@graphter/core";
import { pathConfigStore, nodeRendererStore } from "@graphter/renderer-react";

const treeDataMap: { [key: string]: RecoilValueReadOnly<any> } = {};
const descendentPathDataMap: { [key: string]: RecoilValueReadOnly<Array<Array<PathSegment>>> } = {};

const modelKeySalt = 'c2c87429-dabf-4ea0-b2e1-6e7a6262bc11'
const descendentPathKeySalt = '04ea4750-019b-446d-87f4-c025f837ab7f'

export default {
  getDescendentData: (path: Array<PathSegment>) => {
    const key = `model-${path.join(modelKeySalt)}`
    let treeDataSelector = treeDataMap[key]
    if(treeDataSelector) return treeDataSelector

    treeDataMap[key] = treeDataSelector = selector<any>({
      key: key,
      get: ({ get }) => {
        const config = pathConfigStore.get(path)
        if(!config) throw new Error(`Couldn't find config for node at path '${path.join('/')}'`)
        const renderer = nodeRendererStore.get(config.type)
        if(!renderer) throw new Error(`Couldn't find renderer for type '${config.type}'`)
        return renderer.getRenderedData(path, (path:Array<PathSegment>) => {
          const state = propDataStore.get(path)
          return get(state)
        })
      }
    });

    return treeDataSelector
  },

  getDescendentPaths: (path: Array<PathSegment>): RecoilValueReadOnly<Array<Array<PathSegment>>> => {
    const key = `descendent-paths-${path.join(descendentPathKeySalt)}`
    let descendentPathSelector = descendentPathDataMap[key]
    if(descendentPathSelector) return descendentPathSelector
    descendentPathDataMap[key] = descendentPathSelector = selector<any>({
      key: key,
      get: ({ get }) => {
        const config = pathConfigStore.get(path)
        if(!config) throw new Error(`Couldn't find config for node at path '${path.join('/')}'`)
        const renderer = nodeRendererStore.get(config.type)
        if(!renderer) throw new Error(`Couldn't find renderer for type '${config.type}'`)
        if(!renderer.getPaths) return [ config.id ]
        const paths = renderer.getPaths(path, (path:Array<PathSegment>) => {
          const state = propDataStore.get(path)
          return get(state)
        })
        return paths
      }
    })
    return descendentPathSelector

  }
}