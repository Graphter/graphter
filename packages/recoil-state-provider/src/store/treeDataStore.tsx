import { RecoilValueReadOnly, selector } from "recoil";
import { NodeConfig, PathSegment } from "@graphter/core";
import { nodeRendererStore } from "@graphter/renderer-react";
import { propDataStore } from "./propDataStore";

const treeDataMap: { [key: string]: RecoilValueReadOnly<any> } = {};
const descendentPathDataMap: { [key: string]: RecoilValueReadOnly<Array<Array<PathSegment>>> } = {};

const treeKeySalt = 'c2c87429-dabf-4ea0-b2e1-6e7a6262bc11'
const descendentPathKeySalt = '04ea4750-019b-446d-87f4-c025f837ab7f'

export interface TreeDataStore {
  getDescendentData: (config: NodeConfig, path: Array<PathSegment>) => RecoilValueReadOnly<any>
  getDescendentPaths: (config: NodeConfig, path: Array<PathSegment>) => RecoilValueReadOnly<Array<Array<PathSegment>>>
}

const treeDataStore: TreeDataStore = {
  getDescendentData: (config, path) => {
    if(!config) throw new Error('Config is required to get descendent data')
    if(!path) throw new Error('Path is required to get descendent data')
    const key = `tree-from-${path.join(treeKeySalt)}`
    let treeDataSelector = treeDataMap[key]
    if(treeDataSelector) return treeDataSelector

    treeDataMap[key] = treeDataSelector = selector<any>({
      key: key,
      get: ({ get }) => {
        const getNodeValue = (path:Array<PathSegment>) => {
          const state = propDataStore.get(path)
          return get(state)
        }
        const renderer = nodeRendererStore.get(config.type)
        if(!renderer) throw new Error(`Couldn't find renderer for type '${config.type}'`)

        return renderer.getChildData ?
          renderer.getChildData(config, path, getNodeValue) :
          getNodeValue(path)
      }
    });

    return treeDataSelector
  },

  getDescendentPaths: (config, path): RecoilValueReadOnly<Array<Array<PathSegment>>> => {
    if(!config) throw new Error('Config is required to get descendent data')
    if(!path) throw new Error('Path is required to get descendent data')
    const key = `descendent-paths-${path.join(descendentPathKeySalt)}`
    let descendentPathSelector = descendentPathDataMap[key]
    if(descendentPathSelector) return descendentPathSelector
    descendentPathDataMap[key] = descendentPathSelector = selector<any>({
      key: key,
      get: ({ get }) => {
        const renderer = nodeRendererStore.get(config.type)
        if(!renderer) throw new Error(`Couldn't find renderer for type '${config.type}'`)
        if(!renderer.getChildPaths) return [ [ config.id ] ]
        const childPaths = renderer.getChildPaths(config, path, (path:Array<PathSegment>) => {
          const state = propDataStore.get(path)
          return get(state)
        })
        return [ [config.id], ...childPaths ]
      }
    })
    return descendentPathSelector

  }
}

export default treeDataStore