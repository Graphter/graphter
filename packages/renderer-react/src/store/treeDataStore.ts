/**
 * WARNING: this file is being used directly in various places in the react renderer
 * which is against the principle that the renderer should be agnostic of state management choices.
 * TODO: refactor access to this module through a standardised interface
 */
import { RecoilValueReadOnly, selector } from "recoil";
import propDataStore from "./propDataStore";
import { JsonType, PathSegment } from "@graphter/core";
import * as nodeRendererStore from "./nodeRendererStore";

const treeDataMap: { [key: string]: RecoilValueReadOnly<any> } = {};
const descendentPathDataMap: { [key: string]: RecoilValueReadOnly<Array<Array<PathSegment>>> } = {};

const modelKeySalt = 'c2c87429-dabf-4ea0-b2e1-6e7a6262bc11'
const descendentPathKeySalt = '04ea4750-019b-446d-87f4-c025f837ab7f'

export default {
  getDescendentData: (path: Array<PathSegment>) => {
    const key = `model-${path.join(modelKeySalt)}`
    let treeDataSelector = treeDataMap[key]
    if(treeDataSelector) return treeDataSelector

    treeDataMap[key] = treeDataSelector = selector<any>({
      key: key,
      get: ({ get }) => {
        function getNodeData(path: Array<PathSegment>){
          const nodeConfig = propDataStore.getConfig(path)
          const nodeRenderer = nodeRendererStore.get(nodeConfig.type)
          if(!nodeRenderer) throw new Error(`No renderer for '${nodeConfig.id}' node of type '${nodeConfig.type}'`)
          if(nodeRenderer.jsonType === JsonType.ARRAY){
            if(!nodeConfig.children) throw new Error(`'${nodeConfig.id}' seems to be an array type but has no child config. At least one is required.`)
            const nodeChildStates = propDataStore.getAll(path)
            const childData: Array<any> = nodeChildStates
              .map((childState, i) => {
                if(!childState) throw new Error('Missing state. Should not happen')
                return getNodeData([ ...path, i ])
              })
            return childData
          } else if(nodeRenderer.jsonType === JsonType.OBJECT){
            if(!nodeConfig.children) throw new Error(`'${nodeConfig.id}' seems to be an object type but has no child config. At least one is required.`)
            return nodeConfig.children?.reduce<{ [key: string]: any }>((a, c) => {
              a[c.id] = getNodeData([ ...path, c.id ])
              return a
            }, {})
          } else {
            const nodeState = propDataStore.get(path)
            return get(nodeState)
          }
        }
        return getNodeData(path)
      }
    });

    return treeDataSelector
  },

  getDescendentPaths: (path: Array<PathSegment>): RecoilValueReadOnly<Array<Array<PathSegment>>> => {
    const key = `descendent-paths-${path.join(descendentPathKeySalt)}`
    let descendentPathSelector = descendentPathDataMap[key]
    if(descendentPathSelector) return descendentPathSelector
    descendentPathDataMap[key] = descendentPathSelector = selector<any>({
      key: key,
      get: ({get}) => {
        function recordChildPaths(
          path: Array<PathSegment>,
          descendentPaths: Array<Array<PathSegment>> = []
        ): Array<Array<PathSegment>>{
          descendentPaths.push(path)
          const nodeConfig = propDataStore.getConfig(path)
          const nodeRenderer = nodeRendererStore.get(nodeConfig.type)
          if(!nodeRenderer) throw new Error(`No renderer for '${nodeConfig.id}' node of type '${nodeConfig.type}'`)
          if(nodeRenderer.jsonType === JsonType.ARRAY){
            if(!nodeConfig.children) throw new Error(`'${nodeConfig.id}' seems to be an array type but has no child config. At least one is required.`)
            const nodeChildStates = propDataStore.getAll(path)
            nodeChildStates.forEach((_ , i) => recordChildPaths([...path, i], descendentPaths))
          } else if(nodeRenderer.jsonType === JsonType.OBJECT){
            if(!nodeConfig.children) throw new Error(`'${nodeConfig.id}' seems to be an object type but has no child config. At least one is required.`)
            nodeConfig.children.forEach(childConfig => recordChildPaths([...path, childConfig.id], descendentPaths))
          }
          return descendentPaths
        }
        return recordChildPaths(path)
      }
    })
    return descendentPathSelector

  }
}