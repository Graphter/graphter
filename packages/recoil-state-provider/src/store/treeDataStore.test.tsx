import React from 'react'
import {
  RecoilRoot,
  RecoilValueReadOnly,
  selector,
  useRecoilValue,
  useRecoilValueLoadable
} from "recoil";
import { TreeDataStore } from "./treeDataStore";
import { propDataStore } from "./propDataStore";
import { pathConfigStore, nodeRendererStore } from "@graphter/renderer-react";
import { when } from "jest-when";
import { render } from "@testing-library/react";
import flushPromises from "../../test-utils/flushPromises";

const pathConfigStoreMock = pathConfigStore as jest.Mocked<any>
const nodeRendererStoreMock = nodeRendererStore as jest.Mocked<any>

describe('treeDataStore', () => {
  let treeDataStore: TreeDataStore
  beforeEach(async () => {
    jest.isolateModules(async () => {
      treeDataStore = require('./treeDataStore').default
    })
    jest.resetAllMocks()
    nodeRendererStoreMock.get.mockReset()
  })
  describe('getDescendentData()', () => {
    it('should return a selector that resolves the descendent tree structure', () => {
      function ConsumerComponent() {
        const value = useRecoilValue(treeDataStore.getDescendentData([ 'page' ]))
        return <div>{JSON.stringify(value)}</div>
      }

      when(pathConfigStoreMock.get)
        .calledWith([ 'page' ])
        .mockReturnValueOnce({
          id: 'page',
          type: 'unicorn'
        })
      const getChildDataMock = jest.fn()
      when(getChildDataMock)
        .calledWith([ 'page' ], expect.anything())
        .mockReturnValue('the-child-data')
      when(nodeRendererStoreMock.get)
        .calledWith('unicorn')
        .mockReturnValueOnce({
          getChildData: getChildDataMock
        })
      const {container} = render(
        <RecoilRoot>
          <ConsumerComponent/>
        </RecoilRoot>
      )
      expect(container).toMatchSnapshot()
    })
    it('should skip descendent data resolution if the starting node does not implement a getChildData() function', async () => {
      function ConsumerComponent() {
        const loadableValue = useRecoilValueLoadable(treeDataStore.getDescendentData([ 'page' ]))
        return loadableValue.state === 'hasValue' ?
          <div>{loadableValue.contents}</div> :
          null
      }

      when(pathConfigStoreMock.get)
        .calledWith([ 'page' ])
        .mockReturnValueOnce({
          id: 'page',
          type: 'unicorn'
        })
      when(nodeRendererStoreMock.get)
        .calledWith('unicorn')
        .mockReturnValueOnce({})
      propDataStore.set([ 'page' ], true, 'the-page-value')
      const {container} = render(
        <RecoilRoot>
          <ConsumerComponent/>
        </RecoilRoot>
      )
      await flushPromises()
      expect(container).toMatchSnapshot()
    })
    it('should cache the selector for next time', () => {
      let states: Array<RecoilValueReadOnly<any>> = []

      function ConsumerComponent() {
        const state = treeDataStore.getDescendentData([ 'page' ])
        states.push(state)
        return null
      }

      when(pathConfigStoreMock.get)
        .calledWith([ 'page' ])
        .mockReturnValueOnce({
          id: 'page',
          type: 'unicorn'
        })
      const getChildDataMock = jest.fn()
      when(getChildDataMock)
        .calledWith([ 'page' ], expect.anything())
        .mockReturnValue('the-child-data')
      when(nodeRendererStoreMock.get)
        .calledWith('unicorn')
        .mockReturnValueOnce({
          getChildData: getChildDataMock
        })
      render(
        <RecoilRoot>
          <ConsumerComponent/>
          <ConsumerComponent/>
        </RecoilRoot>
      )
      expect(states.length).toBe(2)
      expect(states[0]).toBe(states[1])
    })
    it('should error if config is missing for the starting node', () => {
      function ConsumerComponent() {
        const value = useRecoilValue(treeDataStore.getDescendentData([ 'page' ]))
        return <div>{JSON.stringify(value)}</div>
      }

      expect(() => render(
        <RecoilRoot>
          <ConsumerComponent/>
        </RecoilRoot>
      )).toThrowErrorMatchingSnapshot()
    })
    it('should error if no renderer is found for the starting node', () => {
      function ConsumerComponent() {
        const value = useRecoilValue(treeDataStore.getDescendentData([ 'page' ]))
        return <div>{JSON.stringify(value)}</div>
      }

      when(pathConfigStoreMock.get)
        .calledWith([ 'page' ])
        .mockReturnValueOnce({
          id: 'page',
          type: 'unicorn'
        })
      expect(() => render(
        <RecoilRoot>
          <ConsumerComponent/>
        </RecoilRoot>
      )).toThrowErrorMatchingSnapshot()
    })
  })
  describe('getDescendentPaths', () => {
    beforeEach(async () => {
      jest.resetAllMocks()
    })
    function ConsumerComponent() {
      const pathsLoadable = useRecoilValueLoadable(treeDataStore.getDescendentPaths([ 'page' ]))
      return pathsLoadable.state === 'hasValue' ?
        <div>
          {JSON.stringify( pathsLoadable.contents)}
        </div> : null
    }
    it('should return a selector that resolves the descendent paths', async () => {
      when(pathConfigStoreMock.get)
        .calledWith([ 'page' ])
        .mockReturnValueOnce({
          id: 'page',
          type: 'unicorn'
        })
      const getChildPathsMock = jest.fn()
      when(getChildPathsMock)
        .calledWith([ 'page' ], expect.anything())
        .mockReturnValueOnce( [['page', 'title'], ['page', 'author']])
      when(nodeRendererStoreMock.get)
        .calledWith('unicorn')
        .mockReturnValueOnce({
          getChildPaths: getChildPathsMock
        })
      const {container} = render(
        <RecoilRoot>
          <ConsumerComponent/>
        </RecoilRoot>
      )
      await flushPromises()
      expect(container).toMatchSnapshot()
    })
    it('should skip descendent path resolution if the starting node does not implement a getChildPaths() function', async () => {
      when(pathConfigStoreMock.get)
        .calledWith([ 'page' ])
        .mockReturnValueOnce({
          id: 'page',
          type: 'unicorn'
        })
      when(nodeRendererStoreMock.get)
        .calledWith('unicorn')
        .mockReturnValueOnce({ })
      const {container} = render(
        <RecoilRoot>
          <ConsumerComponent/>
        </RecoilRoot>
      )
      await flushPromises()
      expect(container).toMatchSnapshot()
    })
    it('should cache the selector for next time', async () => {
      const states: Array<any> = []
      function RecordingConsumerComponent() {
        states.push(treeDataStore.getDescendentPaths([ 'page' ]))
        return null
      }
      when(pathConfigStoreMock.get)
        .calledWith([ 'page' ])
        .mockReturnValueOnce({
          id: 'page',
          type: 'unicorn'
        })
      when(nodeRendererStoreMock.get)
        .calledWith('unicorn')
        .mockReturnValueOnce({ })
      render(
        <RecoilRoot>
          <RecordingConsumerComponent/>
          <RecordingConsumerComponent/>
        </RecoilRoot>
      )
      await flushPromises()
      expect(states.length).toBe(2)
      expect(states[0]).toBe(states[1])
    })
    it('should error if config is missing for the starting node', async () => {
      function ErrorConsumerComponent() {
        const pathsLoadable = useRecoilValueLoadable(treeDataStore.getDescendentPaths([ 'page' ]))
        return <div>
            {JSON.stringify(pathsLoadable)}
          </div>
      }
      const { container } = render(
        <RecoilRoot>
          <ErrorConsumerComponent/>
        </RecoilRoot>
      )
      await flushPromises()
      expect(container).toMatchSnapshot()
    })
    it('should error if no renderer is found for the starting node', async () => {
      function ErrorConsumerComponent() {
        const pathsLoadable = useRecoilValueLoadable(treeDataStore.getDescendentPaths([ 'page' ]))
        return <div>
          {JSON.stringify(pathsLoadable)}
        </div>
      }
      when(pathConfigStoreMock.get)
        .calledWith([ 'page' ])
        .mockReturnValueOnce({
          id: 'page',
          type: 'unicorn'
        })
      const { container } = render(
        <RecoilRoot>
          <ErrorConsumerComponent/>
        </RecoilRoot>
      )
      await flushPromises()
      expect(container).toMatchSnapshot()
    })
  })
})
export {}