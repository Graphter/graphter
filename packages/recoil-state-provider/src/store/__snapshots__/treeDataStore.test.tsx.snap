// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`treeDataStore getDescendentData() should error if no renderer is found for the starting node 1`] = `"Couldn't find renderer for type 'unicorn'"`;

exports[`treeDataStore getDescendentData() should error if null config is passed to the starting node 1`] = `"Config is required to get descendent data"`;

exports[`treeDataStore getDescendentData() should error if undefined config is passed to the starting node 1`] = `"Config is required to get descendent data"`;

exports[`treeDataStore getDescendentData() should return a selector that resolves the descendent tree structure 1`] = `
<div>
  <div>
    "the-child-data"
  </div>
</div>
`;

exports[`treeDataStore getDescendentData() should skip descendent data resolution if the starting node does not implement a getChildData() function 1`] = `
<div>
  <div>
    the-page-value
  </div>
</div>
`;

exports[`treeDataStore getDescendentPaths should error if no renderer is found for the starting node 1`] = `"Couldn't find renderer for type 'unicorn'"`;

exports[`treeDataStore getDescendentPaths should error if null config is passed to the starting node 1`] = `"Config is required to get descendent data"`;

exports[`treeDataStore getDescendentPaths should error if undefined config is passed to the starting node 1`] = `"Config is required to get descendent data"`;

exports[`treeDataStore getDescendentPaths should return a selector that resolves the descendent paths 1`] = `
<div>
  <div>
    [["page"],["page","title"],["page","author"]]
  </div>
</div>
`;

exports[`treeDataStore getDescendentPaths should skip descendent path resolution if the starting node does not implement a getChildPaths() function 1`] = `
<div>
  <div>
    [["page"]]
  </div>
</div>
`;
